### ISSUE TITLE
Rule Change: Allow `declare var` with `no-var`

### ISSUE THREAD

**Issue by aaronadamsCA** [eslint/eslint#15896]
### What rule do you want to change?

no-var

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```ts
declare var analytics: SegmentAnalytics.AnalyticsJS | undefined;
```

### What does the rule currently do for this code?

`Unexpected var, use let or const instead.`

### What will the rule do after it's changed?

Nothing.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

This is essentially a re-opening of #11358, because I've found a scenario where the workaround in that issue is invalid: augmenting a previously declared type.

We use `@types/segment-analytics`, which ships the type `declare var analytics: SegmentAnalytics.AnalyticsJS`. I want to augment this type to possibly be `undefined`, since we want our components to work safely in environments like Storybook where we don't want to run Segment.

Because the shipped definition uses `declare var`, it seems I must also use `declare var` to augment the type. If I try to use `declare const` or `declare let`, I get a broken `any` type.

~~For now it's pretty simple to work around this with an override:~~
This override seemed to suppress the error, but then `--fix` changed `var` to `let` anyway:

```js
    {
      files: ["*.d.ts"],
      rules: {
        "no-var": "off",
      },
    },
```

So now I'm just using a local `// eslint-disable-next-line no-var` until I can find a global solution.

But I think the suggested fix from #11358 is worth considering again, given this use case.

Thanks!

**Comment by mdjermanovic** [comment]
Hi @aaronadamsCA, thanks for the issue!

Core rules do not support TypeScript-specific syntax, so we can't update the core `no-var` rule to recognize `declare var`.

You can ask in https://github.com/typescript-eslint/typescript-eslint repo for an [extension rule](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/eslint-plugin#extension-rules).

Alternatively, you can configure the [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule to disallow `var` but still allow `declare var`:

```json
{
    "parser": "@typescript-eslint/parser",
    "rules": {
        "no-var": "off",
        "no-restricted-syntax": ["error", {
            "selector": "VariableDeclaration[kind='var'][declare!=true]",
            "message": "Unexpected var, use let or const instead."
        }]
    }
}
```

**Comment by aaronadamsCA** [comment]
Ah, that's my mistake, then - I did not understand that from the prior conversation.

Thank you for the help, and for the suggested workaround!
